yesod-test/Yesod/Test.hs

```diff
@@ -24,8 +24,86 @@ This is very useful for testing web applications built in yesod, for example,
 where your forms may have field names generated by the framework or a randomly
 generated CSRF token input.
 
-Your database is also directly available so you can use 'runDB' to set up
+=== Example project
+
+The best way to see an example project using yesod-test is to create a scaffolded Yesod project:
+
+@stack new projectname yesod-sqlite@
+
+(See https://github.com/commercialhaskell/stack-templates/wiki#yesod for the full list of Yesod templates)
+
+The scaffolded project makes your database directly available in tests, so you can use 'runDB' to set up
 backend pre-conditions, or to assert that your session is having the desired effect.
+It also handles wiping your database between each test.
+
+=== Example code
+
+The code below should give you a high-level idea of yesod-test's capabilities.
+Note that it uses helper functions like @withApp@ and @runDB@ from the scaffolded project; these aren't provided by yesod-test.
+
+@
+spec :: Spec
+spec = withApp $ do
+  describe \"Homepage\" $ do
+    it "loads the homepage with a valid status code" $ do
+      'get' HomeR
+      'statusIs' 200
+  describe \"Login Form\" $ do
+    it "Only allows dashboard access after logging in" $ do
+      'get' DashboardR
+      'statusIs' 401
+
+      'get' HomeR
+      -- Assert a \<p\> tag exists on the page
+      'htmlAnyContain' \"p\" \"Login\"
+
+      -- yesod-test provides a 'RequestBuilder' monad for building up HTTP requests
+      'request' $ do
+        -- Lookup the HTML \<label\> with the text Username, and set a POST parameter for that field with the value Felipe
+        'byLabelExact' \"Username\" \"Felipe\"
+        'byLabelExact' \"Password\" "pass\"
+        'setMethod' \"POST\"
+        'setUrl' SignupR
+      'statusIs' 200
+
+      -- The previous request will have stored a session cookie, so we can access the dashboard now
+      'get' DashboardR
+      'statusIs' 200
+
+      -- Assert a user with the name Felipe was added to the database
+      [Entity userId user] <- runDB $ selectList [] []
+      'assertEq' "A single user named Felipe is created" (userUsername user) \"Felipe\"
+  describe \"JSON\" $ do
+    it "Can make requests using JSON, and parse JSON responses" $ do
+      -- Precondition: Create a user with the name \"George\"
+      runDB $ insert_ $ User \"George\" "pass"
+
+      'request' $ do
+        -- Use the Aeson library to send JSON to the server
+        'setRequestBody' ('Data.Aeson.encode' $ LoginRequest \"George\" "pass")
+        'addRequestHeader' (\"Accept\", "application/json")
+        'addRequestHeader' ("Content-Type", "application/json")
+        'setUrl' LoginR
+      'statusIs' 200
+
+      -- Parse the request's response as JSON
+      (signupResponse :: SignupResponse) <- 'requireJSONResponse'
+@
+
+=== HUnit / HSpec integration
```

Comment:
I think this is worth surfacing, to remind people they can use hspec libraries like `hspec-expectations`.

This would be a good place to plug https://hackage.haskell.org/package/hspec-expectations-lifted, but I'm a little hesitant as it hasn't seen a commit in 3 years